analysis/src/python/data_analysis/analysis/statistics.py

```diff
@@ -0,0 +1,241 @@
+
+    """
+
+    issues_stats = {
+        Stats.ISSUE.value: df_issues[IssuesColumns.TYPE.value if by_type else IssuesColumns.CLASS.value].values,
+        Stats.COUNT.value: [df[issue_class].sum() for issue_class in df_issues[IssuesColumns.CLASS.value].values],
+    }
+
+    attr = get_attr(attr)
+
+    for value in attr.values:
+        df_value = df[df[attr.name] == value]
+        issues_stats[value] = []
+
+        for issue_class in df_issues[IssuesColumns.CLASS.value].values:
+            issues_stats[value].append(df_value[issue_class].sum())
+
+    df_stats = pd.DataFrame.from_dict(issues_stats)
+    df_stats = df_stats.groupby([Stats.ISSUE.value], as_index=False).sum()
+    df_stats = df_stats.sort_values(Stats.COUNT.value, ascending=False)
+
+    for value in attr.values:
+        df_stats[value] = df_stats[value] / df[df[attr.name] == value].shape[0]
+    return df_stats
+
+
+def get_client_stats(df: pd.DataFrame) -> pd.DataFrame:
+    """ For each client series e.x. [web, idea, web] count number of submissions series. """
+
+    df_client_stats = df[[SubmissionColumns.GROUP.value, SubmissionColumns.TOTAL_ATTEMPTS.value,
+                          SubmissionColumns.CLIENT_SERIES.value]] \
+        .groupby([SubmissionColumns.CLIENT_SERIES.value, SubmissionColumns.TOTAL_ATTEMPTS.value]) \
+        .count().reset_index()
+    df_client_stats.columns = \
+        [SubmissionColumns.CLIENT_SERIES.value, SubmissionColumns.TOTAL_ATTEMPTS.value, Stats.COUNT.value]
+    df_client_stats.sort_values(Stats.COUNT.value, ascending=False, inplace=True)
+
+    return df_client_stats
+
+
+def filter_by_attempts(df: pd.DataFrame, max_attempts: int, exact_attempts: bool) -> pd.DataFrame:
+    if exact_attempts:
+        return df[df[SubmissionColumns.TOTAL_ATTEMPTS.value] == max_attempts]
+    else:
+        return df[df[SubmissionColumns.TOTAL_ATTEMPTS.value] <= max_attempts]
+
+
+def get_submissions_series_dynamic_by_feature(df: pd.DataFrame, feature: str, attr: AttrType,
+                                              max_attempts: int = 3, exact_attempts: bool = True, is_mean=True):
+    """
+    For attempt count feature value change is submissions which divided by attribute values.
+    Submissions can be filter to analyze submissions with less then `max_attempts` total attempts or
+    exact `max_attempts` if exact_attempts=True.
+    """
+
+    series_stats_dict = {
+        SubmissionColumns.ATTEMPT.value: [],
+        Stats.COUNT.value: [],
+    }
+
+    attr = get_attr(attr)
+    df = filter_by_attempts(df, max_attempts, exact_attempts)
+
+    for value in attr.values:
+        series_stats_dict[value] = []
+
+    for attempt in sorted(df[SubmissionColumns.ATTEMPT.value].unique()):
+        df_attempt = df[df[SubmissionColumns.ATTEMPT.value] == attempt]
+        series_stats_dict[SubmissionColumns.ATTEMPT.value].append(attempt)
+        series_stats_dict[Stats.COUNT.value].append(df_attempt.shape[0])
```

Comment:
Can we move it into a function since you use it in several places?
e.g. create a function that has as an argument a function for the different parts and looks like something like this:
```
for attempt in sorted(df[SubmissionColumns.ATTEMPT.value].unique()):
        df_attempt = df[df[SubmissionColumns.ATTEMPT.value] == attempt]
        series_stats_dict[SubmissionColumns.ATTEMPT.value].append(attempt)
        series_stats_dict[Stats.COUNT.value].append(df_attempt.shape[0])

        function_from_argiments(df_attempt, series_stats_dict)
```